AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/FocusElementLine.java

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package com.microsoft.accessibilityinsightsforandroidservice;
+
+import android.graphics.Canvas;
+import android.graphics.Paint;
+import android.graphics.Rect;
+import android.view.View;
+import android.view.accessibility.AccessibilityNodeInfo;
+import java.util.HashMap;
+
+public class FocusElementLine {
+  private static final String TAG = "FocusElementLine";
+  private AccessibilityNodeInfo eventSource;
+  private AccessibilityNodeInfo previousEventSource;
+  private int yOffset;
+  private int x1Coordinate;
+  private int y1Coordinate;
+  private int x2Coordinate;
+  private int y2Coordinate;
+  private HashMap<String, Paint> paints;
+  private Rect currentRect;
+  private Rect prevRect;
+  private View view;
+
+  public FocusElementLine(
+      AccessibilityNodeInfo eventSource,
+      AccessibilityNodeInfo previousEventSource,
+      HashMap<String, Paint> Paints,
+      View view) {
+    this.view = view;
+    this.eventSource = eventSource;
+    this.previousEventSource = previousEventSource;
+    this.yOffset = setYOffset();
+    this.paints = Paints;
+    this.currentRect = new Rect();
+    this.prevRect = new Rect();
+    this.setCoordinates();
+  }
+
+  public void drawLine(Canvas canvas) {
+    this.drawConnectingLine(
+        this.x1Coordinate,
+        this.y1Coordinate,
+        this.x2Coordinate,
+        this.y2Coordinate,
+        this.paints.get("line"),
+        canvas);
+  }
+
+  private void setCoordinates() {
+    if (this.eventSource == null || this.previousEventSource == null) {
+      return;
+    }
+
+    this.eventSource.getBoundsInScreen(this.currentRect);
+    this.currentRect.offset(0, this.yOffset);
+
+    this.previousEventSource.getBoundsInScreen(this.prevRect);
+    this.prevRect.offset(0, this.yOffset);
+
+    this.x1Coordinate = currentRect.centerX();
+    this.y1Coordinate = currentRect.centerY();
+    this.x2Coordinate = prevRect.centerX();
+    this.y2Coordinate = prevRect.centerY();
+  }
+
+  private int setYOffset() {
```

Comment:
`FocusElementLine.setYOffset` is identical to `FocusElementHighlight.setYOffset`. Would it make sense to make this a utility function that is used by both classes?

Reply:
@DaveTryon I like that idea.  Would you suggest a `FocusElementUtilities` class to handle that call, that is then passed into the constructor of each the `FocusElementLine` and `FocusElementHighlight` classes?

Reply:
@brocktaylor7, the approach you describe makes the most sense when we're changing state within a method. The cleanest approach is usually to allow the injected class to own the state, then the containing class reads the state from the injected class. In *this specific case*, there's no state to manage--we can just pass in a view and get back an offset. We can have a simple static helper class something like this:
```
class OffsetHelper {
  public int getYOffset(View view) {
...
  }
}
```

If the helper method has external dependencies (like the call to `getResources().getDimensionPixelSize()`) that make testing difficult, there may be a benefit to use an injected interface instead of a static class. Ultimately, we want to find the "sweet spot" between too little and too much complexity.

Reply:
Went with a static helper class. PowerMock allows for mocking the system dependencies as well so we're able to test any non-system functionality